sql/parser/builtins.go

```diff
@@ -72,6 +75,74 @@ var builtins = map[string]builtin{
 	"upper": stringBuiltin(func(s string) (Datum, error) {
 		return DString(strings.ToUpper(s)), nil
 	}),
+
+	// TODO(XisiHuang): support the substring(str FROM x FOR y) syntax.
+	"substring": {
+		nArgs: -1,
+		fn: func(args DTuple) (Datum, error) {
+			argsNum := len(args)
+			if argsNum != 2 && argsNum != 3 {
+				return DNull, fmt.Errorf("incorrect number of arguments: %d vs %s",
+					argsNum, "2 or 3")
+			}
```

Comment:
Perhaps `builtin` should have `minArgs` and `maxArgs` fields instead of `nArgs`. Not for this PR, just thinking out loud here.